src/main/java/sparksoniq/spark/ml/AnnotateFunctionIterator.java

```diff
@@ -0,0 +1,67 @@
+package sparksoniq.spark.ml;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.rumbledb.api.Item;
+import org.rumbledb.exceptions.ExceptionMetadata;
+import org.rumbledb.exceptions.UnexpectedTypeException;
+import org.rumbledb.items.ObjectItem;
+import org.rumbledb.runtime.DataFrameRuntimeIterator;
+import org.rumbledb.runtime.RuntimeIterator;
+import sparksoniq.jsoniq.ExecutionMode;
+import sparksoniq.semantics.DynamicContext;
+import sparksoniq.spark.DataFrameUtils;
+
+import java.util.List;
+
+public class AnnotateFunctionIterator extends DataFrameRuntimeIterator {
+
+    private static final long serialVersionUID = 1L;
+
+    public AnnotateFunctionIterator(
+            List<RuntimeIterator> arguments,
+            ExecutionMode executionMode,
+            ExceptionMetadata iteratorMetadata
+    ) {
+        super(arguments, executionMode, iteratorMetadata);
+    }
+
+    @Override
+    public Dataset<Row> getDataFrame(DynamicContext context) {
+        RuntimeIterator inputDataIterator = this.children.get(0);
+        RuntimeIterator schemaIterator = this.children.get(1);
+        ObjectItem schemaItem;
+
+        // materialize singleton pattern
+        schemaIterator.open(context);
+        if (!schemaIterator.hasNext()) {
+            throw new UnexpectedTypeException(
+                    "Schema provided to annotate function can not be an empty sequence.",
+                    getMetadata()
+            );
+        }
+        schemaItem = (ObjectItem) schemaIterator.next();
+        if (schemaIterator.hasNext()) {
+            throw new UnexpectedTypeException(
+                    "Schema provided to annotate function must be a singleton object.",
+                    getMetadata()
+            );
+        }
+        schemaIterator.close();
+
+        if (inputDataIterator.isDataFrame()) {
+            // TODO: alternatively, throw an 'IsAlreadyDataFrame" exception
+            return inputDataIterator.getDataFrame(context); // return unchanged
+        }
```

Comment:
@ghislainfourny how should the desired behavior for annotate() be when called on an existing dataframe?

Reply:
I would say we simply check that the types from the second parameter match the dataframe schema types (which should be easy), and if the check succeeds, return the dataframe as is.

Reply:
That makes sense. If you don't mind it, I would like to merge this PR as is and address this type-checking in a separate PR. I believe some of the utilities I have been implementing for the vectorization of features PR can be re-used for this scheduled improvement.